### PR TITLE
Add typing file for typescript

### PR DESCRIPTION
**Pull request by Vincent-Pang**
I have created a typing file (.d.ts) for typescript user.

Usage:

```
import * as CanCan from 'cancan';

const cancan = new CanCan();

cancan.allow(...);
cancan.can(...);
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -44,7 +44,7 @@ class CanCan {
 		});
 	}
 
-	can(performer, action, target) {
+	can(performer, action, target, payload) {
```
Comment (vadimdemedes) [inline index.js]: Doesn't seem related to that PR.

File: package.json [hunk 2]
```diff
@@ -5,14 +5,16 @@
   "author": "Vadim Demedes <vdemedes@gmail.com>",
   "license": "MIT",
   "repository": "vadimdemedes/cancan",
+  "main": "./index.js",
```
Comment (vadimdemedes) [inline package.json]: Remove that.

File: package.json [hunk 3]
```diff
@@ -5,14 +5,16 @@
...
+  "types": "./index.d.ts",
```
Comment (vadimdemedes) [inline package.json]: I don't think relative path is needed.

File: package.json [hunk 4]
```diff
@@ -5,14 +5,16 @@
...
   "engines": {
     "node": ">= 4"
   },
   "scripts": {
     "test": "xo && ava"
   },
   "files": [
-    "index.js"
+    "index.js", "index.d.ts"
```
Comment (vadimdemedes) [inline package.json]: Each file should be on a new line.

**Comment by Vincent-Pang** [comment]
Great, I have updated according to your comments. Please have a look.

**Comment by vadimdemedes** [comment]
This https://github.com/vadimdemedes/cancan/pull/31#discussion_r146104111 is still true. Also:

> Is there anyone that can confirm that your typings are valid? I'm not familiar with TS.

**Comment by goldenbearkin** [comment]
The type file works great to me. please kindly merge the PR.

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
+declare namespace CanCan {
+    export interface Option {
+        instanceOf?: <T, U>(instance: T, model: U) => boolean;
```
Comment (vadimdemedes) [inline index.d.ts]: I'm not familiar with TS, but where are those T, U declared? Can they be explanatory?

Comment (Vincent-Pang) [reply index.d.ts]: They are called generics, which is a technique to due with multiple types.

For example,
```
// this function accepts an arg and it must be an number.
// what about if I also want the function to accept string, boolean, etc?
function identity(arg: number): number {
    return arg;
}
```

```
// I could declare the type to any, which means it accepts any types. 
// however we are losing type information when using any.
function identity(arg: any): any {
    return arg;
}
```

```
// actually, we can use generic type to capture the type information
function identity<T>(arg: T): T {
    return arg;
}
```

A long explanation can be found in typescript handbook
https://www.typescriptlang.org/docs/handbook/generics.html

**Comment by Vincent-Pang** [comment]
I am going to make a pull request to [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/22079), which is a repo for storing typing files. So that people could `yarn add @types/cancan` if they need the typing file.

As a result, this pull request is no longer needed.

Thank you all.

**Comment by vadimdemedes** [comment]
This is great, thanks!
